src/storage/nats_kv.rs

```diff
@@ -0,0 +1,187 @@
+//! Storage engine backed by NATS Kv
+//!
+//! This storage engine enables storing any WADM related information using NATS Kv as a backend
+
+use std::env;
+use std::sync::Arc;
+use std::sync::RwLock;
+use async_trait::async_trait;
+
+use async_nats::{
+    Client as NatsClient,
+    Error as NatsError,
+    Subscriber,
+    jetstream::kv::Config as JetstreamKvConfig,
+    jetstream::kv::Store as JetstreamKvStore,
+};
+
+use semver::{Version, Prerelease, BuildMetadata};
+
+use super::{
+    EngineMetadata,
+    LatticeStorage,
+    StorageEngine,
+    StorageEngineError,
+    StorageError,
+};
+
+use super::state::{
+    LatticeState,
+    LatticeStateBuilder,
+};
+
+const STORAGE_ENGINE_NAME: &'static str = "nats_kv";
+const STORAGE_ENGINE_VERSION: Version = Version {
+    major: 0,
+    minor: 1,
+    patch: 0,
+    pre: Prerelease::EMPTY,
+    build: BuildMetadata::EMPTY,
+};
+
+use crate::storage::Stores;
+
+const ENV_VAR_STORAGE_NATS_KV_PREFIX: &'static str = "STORAGE_NATS_KV_PREFIX";
+const DEFAULT_NATS_KV_PREFIX: &'static str = "storage";
+
+#[derive(Debug)]
+struct NatsKvStorageEngine {
+    /// URL to the nats instance
+    nats_url: String,
+
+    /// nats prefix for the
+    nats_prefix: String,
+
+    // TODO(REFACTOR): Add code level config object, buildable from json::Value?
+
+    /// NATS client
+    client: Option<NatsClient>,
+
+    /// NATS Jetstream KV Store
+    kv_store: Option<JetstreamKvStore>,
+}
+
+impl NatsKvStorageEngine {
+    fn new(nats_url: String) -> NatsKvStorageEngine {
+        NatsKvStorageEngine {
+            nats_url,
+            nats_prefix: env::var(ENV_VAR_STORAGE_NATS_KV_PREFIX).unwrap_or(String::from(DEFAULT_NATS_KV_PREFIX)),
```

Comment:
These should be config items passed in to the struct rather than env vars here. I just merged #60 that has the flags and stuff set up to get the data!

Reply:
Yup! agreed

- [ ] use config object
- [ ] integrate changes from #60, add flags as necessary